### PR TITLE
feat(v2): allow line highlighting

### PR DESCRIPTION
**Pull request by lex111**
## Motivation

In the documentation, it is sometimes useful to highlight individual lines in blocks of code (see the [example](https://reactjs.org/docs/conditional-rendering.html#inline-if-with-logical--operator) from React).

This PR adds this feature, see the usage example below.
```
```js {1,3-5}
function foo() {
  return bar()
    .then(res => {
      return res.doSomething()
    })
}
```

Lines 1 and between 1 and 2 will be highlighted.

Note: This `{n,n-m}` format de facto used for such purposes.

The only unresolved question is the __default style__, i.e. where (what is the path? theme-classic/../Layout/styles.css?) we need to set CSS styles for the `highlight-line` class so that the user can then override it in own `custom.css` file.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/CodeBlock/index.js [hunk 1]
```diff
@@ -13,7 +13,29 @@ import Clipboard from 'clipboard';
 import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
 import styles from './styles.module.css';
 
-export default ({children, className: languageClassName}) => {
+const regexHighlightRange = /{([\d,-]+)}/;
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/theme/CodeBlock/index.js]: Can we use https://www.npmjs.com/package/parse-numeric-range instead for parsing? Allows for more options and might be more robust than custom regex.

**Comment by endiliey** [comment]
> The only unresolved question is the default style, i.e. where (what is the path? theme-classic/../Layout/styles.css?) we need to set CSS styles for the highlight-line class so that the user can then override it in own custom.css file.

Regarding the css, we can do it like gatsby https://www.gatsbyjs.org/packages/gatsby-remark-prismjs/#optional-add-line-highlighting-styles. They have to create their own css there

> Do you think it makes sense to add this feature to the Live Codeblock as well? I'm leaning towards no but would just like to bring this up.
Technically yes because we'll want to demo it on our docs as well. >.<

**Comment by endiliey** [comment]
P.S: Documentation needed

File: packages/docusaurus-theme-classic/src/theme/CodeBlock/index.js (hunk 1 above)
Comment (lex111) [reply packages/docusaurus-theme-classic/src/theme/CodeBlock/index.js]: @yangshun done :heavy_check_mark:

**Comment by lex111** [comment]
> Regarding the css, we can do it like gatsby...

There is the `.highlight-line` class, I need to set default CSS rules for it, in which CSS file to do this?

> P.S: Documentation needed

Can you help with this?

**Comment by yangshun** [comment]
@lex111 thanks for this PR, highly appreciated. As for docs, you could mention this new syntax in this section - https://v2.docusaurus.io/docs/markdown-features#syntax-highlighting

**Comment by lex111** [comment]
@yangshun I thought it would be faster if you document this feature, especially since I am not familiar with the guidelines that have been adopted, and we will only lose time (review-changes-review).
...And I still have sites that are not finished for migration, and I have a headache with redirects (sorry to mention this again!).

**Comment by yangshun** [comment]
No problem, I can take over that. As for redirects, I think we should handle it in Docusaurus side as a first-class feature since many sites will need it. @endiliey let's find time to work on that?

**Comment by lex111** [comment]
Conflicts resolved.

I thought to look into the source code v1, there seems to be a file generated with the `Redirect` component. I don’t know how to do it in the second version... but time is running out! I still wanted to complete migration of requested sites.

**Comment by endiliey** [comment]
There is no way to redirect in v1. That redirect component is only used to redirect docs/id to docs/en/id if language is enabled.

Actually there’s a way to do it right now in v2 that I think is a very good solution to redirect. 

The key is using `postBuild` plugin hook to create a HTML redirect. User can define `fromPath` and `toPath`.
Maybe something like

```json
{
    "redirects": [
            {
                "from": "oldpage.html",
                "to": "newpage.html"
            },
            {
                "from": "olddir/oldpage.html",
                "to": "newdir/newpage.html"
            },
        ]
}

```
The HTML redirect can look like this
```html
<!DOCTYPE HTML>
<html lang="en-US">
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=docs/id-of-doc-to-land-on.html">
    <script type="text/javascript">
      window.location.href = 'docs/id-of-doc-to-land-on.html';
    </script>
    <title>Your Site Title Here</title>
  </head>
  <body>
    If you are not redirected automatically, follow this <a href="docs/id-of-doc-to-land-on.html">link</a>.
  </body>
</html>
```

We shouldn’t do client side redirect because not every static hosting service support 404 page or fallback to index.html. Some host like netlify is able to do server side redirect, but not all can do it.

Sent from Mail for Windows 10

**Comment by lex111** [comment]
> There is no way to redirect in v1. That redirect component is only used to redirect docs/id to docs/en/id if language is enabled.

That was an assumption, I'm not familiar with the internals Docusaurus. I thought you could do something similar in v2...

> There is no way to redirect in v1. That redirect component is only used to redirect docs/id to docs/en/id if language is enabled.

Wow, although I don’t like the need to manually set urls.. but OK, how do I programmatically get a list of all pages?

> We shouldn’t do client side redirect because not every static hosting service support 404 page or fallback to index.html. Some host like netlify is able to do server side redirect, but not all can do it.

I completely agree, so I did not consider such an option.

**Comment by endiliey** [comment]
Can do it like sitemap
https://github.com/facebook/docusaurus/blob/master/packages/docusaurus-plugin-sitemap/src/index.js

Sent from Mail for Windows 10

**Comment by endiliey** [comment]
> There is the .highlight-line class, I need to set default CSS rules for it, in which CSS file to do this?
On each co-located styles.css. Let's just convert this css module to traditional css for easy override

eg: https://github.com/facebook/docusaurus/blob/2b0cc9bc32711a02d38255160a1fffdcc31bdceb/packages/docusaurus-theme-classic/src/theme/CodeBlock/styles.module.css

wdyt @yangshun

**Comment by lex111** [comment]
This is the CSS Module file, I meant the regular one.

**Comment by endiliey** [comment]
Hmm i was thinking to rename the css modules file to just a normal css file. There is no global css file in v2 like v1 (except the infima css).

**Comment by yangshun** [comment]
I will patch this PR when I get the time (probably this weekend or early next week).

**Comment by yangshun** [comment]
Btw @lex111 we would like to send you a Docusaurus plushie for your great contributions. If you're interested would you PM me your mailing address on Discord? Thanks!

**Comment by endiliey** [comment]
I vouch for the plushie for lex111. Awesome job

**Comment by yangshun** [comment]
I've resolved the conflicts. I want to add the docs for this feature but I can only do so after using this feature in our docs and adding some highlight classes (so I know what docs to write).

Hence I will merge this PR first and make another PR to make the docs and CSS updates.

**Comment by lex111** [comment]
@yangshun thank you!

**Comment by yangshun** [comment]
@lex111 I should thank you for being such a rockstar 🎸

**Comment by bravo-kernel** [comment]
👏

**Pull request merged**